versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -0,0 +1,936 @@
+   * @throws ReferenceNotFoundException if either {@code ref} does not exist or {@code hashOnRef}
+   *     does not exist on {@code ref}
+   */
+  protected Hash hashOnRef(
+      OP_CONTEXT ctx,
+      Hash knownHead,
+      NamedRef ref,
+      Optional<Hash> hashOnRef,
+      Consumer<CommitLogEntry> commitLogVisitor)
+      throws ReferenceNotFoundException {
+    if (hashOnRef.isPresent()) {
+      Hash suspect = hashOnRef.get();
+      if (suspect.equals(NO_ANCESTOR)) {
+        // If the client requests 'NO_ANCESTOR' (== beginning of time), skip the existence-check.
+        if (commitLogVisitor != null) {
+          commitLogFetcher(ctx, knownHead).forEach(commitLogVisitor);
+        }
+        return suspect;
+      }
+
+      Stream<Hash> hashes;
+      if (commitLogVisitor != null) {
+        hashes =
+            commitLogFetcher(ctx, knownHead).peek(commitLogVisitor).map(CommitLogEntry::getHash);
+      } else {
+        hashes = commitLogHashFetcher(ctx, knownHead);
+      }
+      if (hashes.noneMatch(suspect::equals)) {
+        throw hashNotFound(ref, suspect);
+      }
+      return suspect;
+    } else {
+      return knownHead;
+    }
+  }
+
+  /** Load the commit-log entry for the given hash, return {@code null}, if not found. */
+  protected abstract CommitLogEntry fetchFromCommitLog(OP_CONTEXT ctx, Hash hash);
+
+  /**
+   * Fetch multiple {@link CommitLogEntry commit-log-entries} from the commit-log. The returned list
+   * must have exactly as many elements as in the parameter {@code hashes}. Non-existing hashes are
+   * returned as {@code null}.
+   */
+  protected abstract List<CommitLogEntry> fetchPageFromCommitLog(OP_CONTEXT ctx, List<Hash> hashes);
+
+  /** Reads from the commit-log starting at the given commit-log-hash. */
+  protected Stream<CommitLogEntry> commitLogFetcher(OP_CONTEXT ctx, Hash initialHash) {
+    return logFetcher(ctx, initialHash, this::fetchPageFromCommitLog, CommitLogEntry::getParents);
+  }
+
+  /**
+   * Like {@link #commitLogFetcher(Object, Hash)}, but only returns the {@link Hash commit-log-entry
+   * hashes}, which can be taken from {@link CommitLogEntry#getParents()}, thus no need to perform a
+   * read-operation against every hash.
+   */
+  private Stream<Hash> commitLogHashFetcher(OP_CONTEXT ctx, Hash initialHash) {
+    return logFetcher(
+        ctx,
+        initialHash,
+        (c, hashes) -> hashes,
+        hash -> {
+          CommitLogEntry entry = fetchFromCommitLog(ctx, hash);
+          if (entry == null) {
+            return Collections.emptyList();
+          }
+          return entry.getParents();
+        });
+  }
+
+  /**
+   * Constructs a {@link Stream} of entries for either the global-state-log or a commit-log. Use
+   * {@link #commitLogFetcher(Object, Hash)} or the similar implementation for the global-log for
+   * non-transactional adapters.
+   */
+  protected <T> Stream<T> logFetcher(
```

Comment:
the javadoc here implies that the global-state has a history? Should it? Specifically to the iceberg case the global state of  a table becomes invalid once new global state is written. Just looking for clarification and to ensure that we can't time travel on teh global state

Reply:
Right. There's only one global-state for any contents-id.
For no-sql databases it is implemented as a log-structure for technical reasons to have just one CAS operation.
For sql databases it just a table with contents-id + type as the primary key.

Reply:
Note: global-state is not exposed as a log or history or the ability for time-travel.